core/src/main/java/jenkins/agents/WebSocketAgents.java

```diff
@@ -149,22 +150,12 @@ protected void error(Throwable cause) {
             LOGGER.log(Level.WARNING, null, cause);
         }
 
-        class Transport extends AbstractByteArrayCommandTransport {
+        class Transport extends AbstractByteBufferCommandTransport {
 
             @Override
-            public void setup(AbstractByteArrayCommandTransport.ByteArrayReceiver bar) {
-                receiver = bar;
-            }
-
-            @Override
-            public void writeBlock(Channel chnl, byte[] bytes) throws IOException {
-                LOGGER.finest(() -> "writing block of length " + bytes.length + " to " + agent);
-                try {
-                    sendBinary(ByteBuffer.wrap(bytes)).get();
-                } catch (Exception x) {
-                    x.printStackTrace();
-                    throw new IOException(x);
-                }
+            protected void write(ByteBuffer header, ByteBuffer data) throws IOException {
+                sendBinary(header);
+                sendBinary(data);
```

Comment:
```suggestion
                sendBinary(header, false);
                sendBinary(data, true);
```

if `WebSocketSession` is extended in the obvious way